IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m

```diff
@@ -217,4 +221,161 @@ For asymmetric keys this holds the public key hash which allows digital identity
+    
+    // If tenantId is missing, the caller may have requested for primary registration. Query keychain to get the primary registration tenantId
+    if (tenantId == nil)
+    {
+        NSError *error;
+        NSString *primaryRegTenantId = [MSIDWorkPlaceJoinUtilBase getPrimaryEccTenantWithSharedAccessGroup:defaultSharedAccessGroup context:context error:&error];
+        if (!primaryRegTenantId)
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, error.description, error.code);
+            // If tenantId for default primary registration is not found, default registration should have a tenantId associated with it, fast returning here.
+            return nil;
+        }
+        tenantId = primaryRegTenantId;
+    }
+   
+    tag = [NSString stringWithFormat:@"%@#%@", kWPJPrivateKeyIdentifier, tenantId];
+    tagData = [tag dataUsingEncoding:NSUTF8StringEncoding];
+    // Default registrations can be done using RSA/ECC in iOS and only ECC in macOS. 1st Looking for RSA device key in the keychain.
+    __unused NSDictionary *extraDefaultPrivateKeyAttributes = @{ (__bridge id)kSecAttrApplicationTag : tagData,
+                                                        (__bridge id)kSecAttrAccessGroup : defaultSharedAccessGroup,
+                                                        (__bridge id)kSecAttrKeyType : (__bridge id)kSecAttrKeyTypeRSA };
+    
+    extraCertAttributes = @{ (__bridge id)kSecAttrAccessGroup : defaultSharedAccessGroup };
+    // In macOS, default registrations can only be ECC. Skip checking default RSA registration for macOS.
+#if !TARGET_OS_OSX
+    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, context, @"Checking keychain for default registration done using RSA key.");
+    defaultKeys = [self findWPJRegistrationInfoWithAdditionalPrivateKeyAttributes:extraDefaultPrivateKeyAttributes certAttributes:extraCertAttributes context:context];
+
+    // If secondary Identity was found, return it
+    if (defaultKeys)
+    {
+        defaultKeys.keyChainVersion = MSIDWPJKeychainAccessGroupV2;
+        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, context, @"Returning RSA private device key from default registration.");
+        return defaultKeys;
+    }
+#endif
+    
+    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, context, @"Checking keychain for default registration done using ECC key.");
+    // Since the defualt RSA search returned nil in iOS, the key might be an ECC key. Use the tag specific for EC device key and re-try
+    tag = [NSString stringWithFormat:@"%@%@", tag, kECPrivateKeyTagSuffix];
+    tagData = [tag dataUsingEncoding:NSUTF8StringEncoding];
+    NSMutableDictionary *privateKeyAttributes = [[NSMutableDictionary alloc] initWithDictionary:extraDefaultPrivateKeyAttributes];
+    [privateKeyAttributes setObject:tagData forKey:(__bridge id)kSecAttrApplicationTag];
+    if (@available(macOS 10.15, *))
+    {
+        [privateKeyAttributes setObject:@YES forKey:(__bridge id)kSecUseDataProtectionKeychain];
```

Comment:
Wouldn't this also set this on iOS? (I never remember how if @available checks work exactly). Maybe this should have another #if TARGET_OS_OSX around it? (here and in other places above)

Reply:
Thanks, you're right, added #if TARGET_OS_OSX macro surrounding @available references. Although in this case, setting kSecUseDataProtectionKeychain to YES for iOS will be futile as it is already enabled for iOS for iOS style keychain :)